File: bindings/prolog/safe_crypto.pl
%% SPDX-License-Identifier: PMPL-1.0-or-later
%% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%%
%% Proven SafeCrypto - Extended FFI bindings to libproven crypto operations.
%% All crypto operations are performed in verified Idris 2 code via libproven.

:- module(safe_crypto, [
    constant_time_equals/2,
    bytes_to_hex/2,
    generate_token/2,
    random_int/3
]).

:- use_module(proven_crypto, [
    constant_time_equals/2 as pc_constant_time_equals,
    bytes_to_hex/2 as pc_bytes_to_hex,
    generate_token/2 as pc_generate_token,
    random_int/3 as pc_random_int
]).

%! constant_time_equals(+A, +B) is semidet.
%
%  Constant-time byte comparison via libproven (timing-attack safe).
constant_time_equals(A, B) :-
    pc_constant_time_equals(A, B).

%! bytes_to_hex(+Bytes, -Hex) is det.
%
%  Convert list of bytes to hexadecimal string via libproven.
bytes_to_hex(Bytes, Hex) :-
    pc_bytes_to_hex(Bytes, Hex).

%! generate_token(+Length, -Token) is det.
%
%  Generate a random hexadecimal token via libproven.
generate_token(Length, Token) :-
    pc_generate_token(Length, Token).

%! random_int(+Min, +Max, -Result) is det.
%
%  Generate a random integer in range [Min, Max] via libproven.
random_int(Min, Max, Result) :-
    pc_random_int(Min, Max, Result).

File: bindings/java/src/main/java/io/github/hyperpolymath/proven/SafeCrypto.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
package io.github.hyperpolymath.proven;

import com.sun.jna.Memory;

import java.util.Optional;

/**
 * Safe cryptographic operations via libproven FFI.
 *
 * Provides constant-time comparison and secure random byte generation.
 * All logic delegates to the Idris 2 verified implementation through
 * the Zig C ABI. No cryptographic logic is reimplemented in Java.
 */
public final class SafeCrypto {

    private static final ProvenLibrary LIB = ProvenLibrary.INSTANCE;

    /** Maximum random byte request size (1 MiB). */
    private static final int MAX_RANDOM_BYTES = 1024 * 1024;

    private SafeCrypto() {}

    /**
     * Constant-time byte array comparison to prevent timing attacks.
     *
     * @param a first byte array
     * @param b second byte array
     * @return true if equal, empty if null pointer error
     */
    public static Optional<Boolean> constantTimeEquals(byte[] a, byte[] b) {
        Memory memA = NativeUtil.toNativeBytes(a);
        Memory memB = NativeUtil.toNativeBytes(b);
        if (memA == null || memB == null) {
            return Optional.empty();
        }
        return NativeUtil.extractBool(
            LIB.proven_crypto_constant_time_eq(memA, a.length, memB, b.length));
    }

    /**
     * Generate cryptographically secure random bytes.
     *
     * @param length number of bytes to generate (1 to 1 MiB)
     * @return the random bytes, or empty on invalid length or allocation error
     */
    public static Optional<byte[]> randomBytes(int length) {
        if (length <= 0 || length > MAX_RANDOM_BYTES) {
            return Optional.empty();
        }
        Memory mem = new Memory(length);
        int statusCode = LIB.proven_crypto_random_bytes(mem, length);
        ProvenStatus status = ProvenStatus.fromCode(statusCode);
        if (status.isError()) {
            return Optional.empty();
        }
        return Optional.of(mem.getByteArray(0, length));
    }
}
